Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,266 @@
+        
+        Parameters
+        ----------
+        x : DataContainer (e.g. ImageData object)
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+        function_num: `int` 
+            Between 0 and n-1, where n is the number of functions in the list  
+        Returns
+        --------
+        DataContainer (e.g. ImageData object)
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+    
+
+        self._svrg_iter_number += 1
+
+        self.stoch_grad_at_iterate = self.functions[function_num].gradient(x)
+
+        if self.store_gradients is True:
+            self._stochastic_grad_difference = self.stoch_grad_at_iterate.sapyb(
+                1., self._list_stored_gradients[function_num], -1.)
+        else:
+            self._stochastic_grad_difference = self.stoch_grad_at_iterate.sapyb(
+                1., self.functions[function_num].gradient(self.snapshot), -1.)
+
+        self._update_data_passes_indices([function_num])
+
+        out = self._stochastic_grad_difference.sapyb(
+                self.num_functions, self._full_gradient_at_snapshot, 1., out=out)
+
+        return out
+
+    def _update_full_gradient_and_return(self, x, out=None):
+        """
+        Takes a "snapshot" at the point :math:`x`, saving both the point :math:` \tilde{x}=x` and its gradient :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}`. The function returns :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}` as the gradient calculation. If :code:`store_gradients==True`, the gradient of all the :math:`f_i`s is computed and stored at the "snapshot"..
+        
+        Parameters
+        ----------
+        Takes a "snapshot" at the point :math:`x`. The function returns :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}` as the gradient calculation. If :code:`store_gradients==True`, the gradient of all the :math:`f_i`s is stored, otherwise only  the sum of the gradients and the snapshot point :math:` \tilde{x}=x` are stored.
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+
+        Returns
+        --------
+        DataContainer (e.g. ImageData object)
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+
+        self._svrg_iter_number += 1 
+
+        if self.store_gradients is True:
+            if self._list_stored_gradients is None: 
+                # Save the gradient of each individual f_i and the gradient of the full sum at the point x. 
+                self._list_stored_gradients = [
+                    fi.gradient(x) for fi in self.functions]
+                self._full_gradient_at_snapshot = sum(self._list_stored_gradients, start=0*x)
```

Comment:
This is not for this PR, but maybe useful for later. Apart from gpu/cpu transfer time, this line costs a lot. The reason is that we sequentially compute the gradient of `num_subsets` functions. The same happens when computing the objective. 

1 Gradient computation = 1 direct, 1 adjoint of the ProjectionOperator. It may seem ok for 2D, but for large 3D and not so many subsets compared to the total number of projections, this takes A LOT OF TIME.

One option is `dask`. The second option without adding dask is to create a 3D Projection Operator, `A=ProjectionOperator(ig3D, ag3D)` and the corresponding function, e.g, `f = LeastSquares(A=A, b=b)`. We need some checking about the type of `fis`. 
The `fis` have all the information to create `A` from `Ais` and `b` from `bis`.

Both should give in theory the same `gradient`. In practice, there is some precision error when you take the sum. Basically, for both the gradient and objective computation, I use 
this for large 3D data, which is faster and more accurate.

Reply:
Thanks Vaggelis. I added in an issue to track this discussion #1865